karapace/avro_compatibility.py

```diff
@@ -0,0 +1,268 @@
+        if reader.type == BYTES:
+            if writer.type == STRING:
+                return result
+            return result.merged_with(self.type_mismatch(reader, writer, location))
+        if reader.type == STRING:
+            if writer.type == BYTES:
+                return result
+            return result.merged_with(self.type_mismatch(reader, writer, location))
+        if reader.type in {ARRAY, MAP, FIXED, ENUM, RECORD}:
+            return result.merged_with(self.type_mismatch(reader, writer, location))
+        if reader.type == UNION:
+            reader = cast(UnionSchema, reader)
+            for reader_branch in reader.schemas:
+                compat = self.get_compatibility(reader_branch, writer)
+                if compat.compatibility is SchemaCompatibilityType.compatible:
+                    return result
+            # No branch in reader compatible with writer
+            message = f"reader union lacking writer type {writer.type}"
+            return result.merged_with(
+                SchemaCompatibilityResult.incompatible(SchemaIncompatibilityType.missing_union_branch, message, location)
+            )
+        raise AvroRuntimeException(f"Unknown schema type: {reader.type}")
+
+    # pylint: enable=too-many-return-statements
+
+    @staticmethod
+    def check_schema_names(reader: NamedSchema, writer: NamedSchema, location: List[str]) -> SchemaCompatibilityResult:
+        result = SchemaCompatibilityResult.compatible()
+        location.append("name")
+        if not reader.name == writer.name:
+            message = f"expected: {writer.fullname}"
+            result = SchemaCompatibilityResult.incompatible(SchemaIncompatibilityType.name_mismatch, message, location)
+        location.pop()
+        return result
+
+    @staticmethod
+    def check_fixed_size(reader: FixedSchema, writer: FixedSchema, location: List[str]) -> SchemaCompatibilityResult:
+        result = SchemaCompatibilityResult.compatible()
+        location.append("size")
+        actual = reader.size
+        expected = writer.size
+        if actual != expected:
+            message = f"expected: {expected}, found {actual}"
+            result = SchemaCompatibilityResult.incompatible(
+                SchemaIncompatibilityType.fixed_size_mismatch,
+                message,
+                location,
+            )
+        location.pop()
+        return result
+
+    @staticmethod
+    def check_reader_enum_contains_writer_enum(
+        reader: EnumSchema, writer: EnumSchema, location: List[str]
+    ) -> SchemaCompatibilityResult:
+        result = SchemaCompatibilityResult.compatible()
+        location.append("symbols")
+        writer_symbols, reader_symbols = set(writer.symbols), set(reader.symbols)
+        extra_symbols = writer_symbols.difference(reader_symbols)
+        # TODO-> it looks like this branch is never actually reached
+        if extra_symbols and False:
```

Comment:
TBH, I don't get this piece. `xxx and False` will always be `False`.

Reply:
basically, with the way enums are implemented in python, the entire method will always return a compatible result, which means it can be done with entirely (it's only used in one location). I started writing it, then realised i can't reach that particular branch without defaults, then kind of only ignored the if. On second thought it's probably better if i delete all of it

Reply:
Hm...

```json
{
    "type": "enum",
    "name": "Suit",
    "symbols": ["SPADES", "HEARTS", "DIAMONDS", "CLUBS"],
    "default": "SPADES"
}
```

and

```json
{
    "type": "enum",
    "name": "Suit",
    "symbols": ["SPADES", "HEARTS", "DIAMONDS", "CLUBS"]
}
```

are different, right? `Parse` recognizes the difference (`"default"`).

I can't find the place where an `EnumSchema` turns into a Python enum (or vice versa). Could you please point me to it, maybe we'll figure something out.

Reply:
@ivanyu 
https://github.com/apache/avro/blob/8c15c1241fb52f3ce6eb480bbe16ff70d3bc426a/lang/py3/avro/schema.py#L681-L735

Reply:
As for the actual initialization :
Java does it like so: https://github.com/apache/avro/blob/942ec8e0cc011ec9fcfff7ba9f5c656bb54d51d1/lang/java/avro/src/main/java/org/apache/avro/Schema.java#L1696-L1700
Python like so: https://github.com/apache/avro/blob/8c15c1241fb52f3ce6eb480bbe16ff70d3bc426a/lang/py3/avro/schema.py#L1125-L1128

It seems it would be pretty straight forward to add enum support, but we would have to patch avro